✨ `Protocol` Translation (#1365)
SwiftGGTeam/the-swift-programming-language-in-chinese: swift-6-beta.docc/LanguageGuide/Protocols.md

# 协议

定义所遵循的类型必须实现的方法、属性和其他要求。

> **协议（Protocol）** 定义了满足特定任务或功能所需的方法、属性和其他要求的蓝图。类、结构体或枚举可以 **遵循（adopt）** 该协议，并提供协议要求的具体实现。任何满足协议要求的类型都被称为 **符合（conform）** 该协议。

Author: 这里我其实比较纠结，原文有很多 "adopt and conform to"，之前的翻译就是“采用并遵循”，感觉怪怪的，所以我想换成“遵循并符合”感觉会自然一些。不过单独说 conform 翻译成 遵循 确实也更好理解，可以再讨论讨论。

另外换行已经在群里讨论过了，会有多的空格，先不换行了。

> **协议（Protocol）** 定义了满足特定任务或功能所需的方法、属性和其他要求的蓝图。类、结构体或枚举可以 **遵循（adopt）** 该协议，并提供协议要求的具体实现。任何满足协议要求的类型都被称为 **符合（conform）** 该协议。

Author: 改回来了

  ```
-->

> 如果一个类拥有父类，应该将父类名放在其他遵循的协议名之前，以逗号分隔：

Author: 精确翻译肯定是 任何，只是感觉在中文里不如 其他 更自然。为了精确的话我还是改成 任何 吧

  ```
-->

> 注意: 由于协议是类型，它们的名称应以大写字母开头（如 `FullyNamed` 和 `RandomNumberGenerator`），以与 Swift 中其他类型的命名规范（如 `Int`、`String` 和 `Double`）保持一致。

## 属性要求

> 协议可以要求遵循协议的类型提供特定名称和类型的实例属性或类型属性。协议不指定属性是存储属性还是计算属性，它只指定属性的名称和类型。此外，协议还指定属性是 **可读** 的还是 **可读可写的**。

Author: 好的，这里应该是觉得上一版这样处理还不错，就没有遵照原文。我改改吧

  ```
-->

> ## 变值方法要求

Author: 有什么提议吗

> ## 变值方法要求

Reviewer: 我看之前版本的名称叫做  
异变方法要求
现在看这个翻译的 也没毛病， 不知道就是跟不跟之前的保持统一

> ## 变值方法要求

Author: 这个之前应该群里讨论过，可以少 care 历史版本是怎么样的，更重要的是当前版本的自洽。这里其他地方我翻译的都是变值，这里也就统一了～

  ```
-->

> 如果一个子类重写了父类的指定构造器，并且该构造器满足了某个协议的要求，那么该构造器的实现需要同时标注 `required` 和 `override` 修饰符：

Reviewer 2: ```suggestion
如果一个子类重写了父类的指定构造器，并且该构造器满足了某个协议的构造器要求，那么该构造器的实现需要同时标注 `required` 和 `override` 修饰符：
```

  ```
-->

## 协议作为类型

协议本身并不实现任何功能。尽管如此，你仍然可以在代码中将协议用作类型。

最常见的将协议用作类型的方式是将其用作泛型约束。具有泛型约束的代码可以与任何符合该协议的类型一起工作，具体的类型由使用该 API 的代码选择。例如，当你调用一个函数并传入一个参数，而该参数的类型是泛型时，调用者会选择具体的类型。

在代码中使用不透明类型时，可以与某个符合该协议的类型一起工作。底层类型在编译时是已知的，API 实现会选择该类型，但该类型的身份对 API 的使用方是隐藏的。使用不透明类型可以防止 API 的实现细节泄露到抽象层之外 —— 例如，通过隐藏函数的具体返回类型，并仅保证该值符合给定的协议。

> 代码使用装箱（boxed）的协议类型时，可以与任何在运行时选择的、符合该协议的类型一起工作。为了支持这种运行时的灵活性，Swift 在必要时会添加一个间接层 —— 称为 **箱子（box）**，这会带来性能开销。由于这种灵活性，Swift 在编译时无法知道底层类型，这意味着你只能访问协议所要求的成员。要访问底层类型的任何其他 API，都需要在运行时进行强制转换。

Reviewer 2: ```suggestion
代码使用装箱（boxed）的协议类型时，可以与任何在运行时选择的、符合该协议的类型一起工作。为了支持这种运行时的灵活性，Swift 在必要时会添加一个间接层 —— 称为 **箱子（box）**，这会带来性能开销。由于这种灵活性，Swift 在编译时无法知道底层类型，这意味着你只能访问协议所要求的成员。要访问底层类型的任何其他 API，都需要在运行时进行类型转换。
```

that cannot be optimized away.
-->

> ## 代理

Author: 嗯嗯，群里也发起过投票，结果是 代理6-0委托

  ```
-->

`beginConcert(in:)` 函数接受一个类型为 `Location & Named` 的参数，这意味着“任何 `Location` 的子类，并且遵循 `Named` 协议”。在这个例子中，`City` 就满足这样的条件。

将 `birthdayPerson` 传入 `beginConcert(in:)` 函数是不合法的，因为 `Person` 不是 `Location` 的子类。同理，如果你新建一个类继承于 `Location`，但是没有遵循 `Named` 协议，用这个类的实例去调用 `beginConcert(in:)` 函数也是不合法的。

> ## 检查协议一致性

Reviewer 2: ```suggestion
## 检查是否遵循协议
```
会不会更好点？

> 你可以使用 <doc:TypeCasting> 中描述的 `is` 和 `as` 操作符来检查协议一致性，即是否遵循某协议，并且可以转换到指定的协议类型。检查和转换协议的语法与检查和转换类型是完全一样的：

Reviewer 2: ```suggestion
你可以使用 <doc:TypeCasting> 中描述的 `is` 和 `as` 操作符来检查是否遵循某协议，并且可以类型转换到指定的协议。检查和转换协议的语法与检查和转换类型是完全一样的：
```
